src/harvest.js

```
@@ -0,0 +1,213 @@
+			throw (e);
+		}
+
+		if (!onlyOnce) {
+			logger.log('debug', `Waiting ${pollInterval / 1000} seconds before polling again`);
+			await setTimeoutPromise(pollInterval);
+			writePollChangeTimestamp(timeBeforeFetching);
+			return process({pollChangeTime: timeBeforeFetching.add(1, 'seconds')});
+		}
+
+		function getPollChangeTime() {
+			if (fs.existsSync(changeTimestampFile)) {
+				const data = JSON.parse(fs.readFileSync(changeTimestampFile, 'utf8'));
+				return moment(data.timestamp);
+			}
+
+			if (pollChangeTimestamp) {
+				return moment(pollChangeTimestamp);
+			}
+
+			return moment();
+		}
+
+		function writePollChangeTimestamp(time) {
+			fs.writeFileSync(changeTimestampFile, JSON.stringify({
+				timestamp: time.format()
+			}));
+		}
+
+		// ListRecords can only fetch 100 records at the time.
+		// Each cycle fetches max 100 records, filters records, passes them to callback and passes possible resumption token to new cycle
+		async function harvest(token) {
+			const url = new URL(harvestURL);
+
+			if (token) {
+				url.search = new URLSearchParams({
+					verb: 'ListRecords',
+					resumptionToken: token
+				});
+			} else {
+				url.search = new URLSearchParams({
+					verb: 'ListRecords',
+					from: getPollChangeTime().utc().format(),
+					metadataPrefix: harvestMetadata
+				});
+				originalUrl = url.toString();
+			}
+
+			let response = null;
+			try {
+				response = await fetch(url.toString());
+			} catch (e) {
+				logger.log('warn', `Query failed: ${e}`);
+				failedQueries++;
+				if (failedQueries >= 5) {
+					throw new Error(JSON.stringify({time: moment(), query: url.toString(), queryOriginal: originalUrl, originalError: e}, null, 2));
+				}
+
+				return harvest(token || '');
+			}
+
+			failedQueries = 0;
+
+			if (response.status === HttpStatusCodes.OK) {
```

Split the block into functions so the code becomes easier to read.

Also `404` is not acceptable and should throw an error.

---

This is copied from Helmet, 404 else if block there too; https://github.com/NatLibFi/melinda-record-import-harvester-helmet/blob/master/src/harvest.js line 140